northstar/src/runtime/state.rs

```diff
@@ -109,615 +89,464 @@ impl ProcessContext {
         } else {
-            if self.applications.get(&name).is_some() {
-                return Err(Error::ContainerAlreadyInstalled(name));
-            }
-            let app = Application::new(container);
-            self.applications.insert(name, app);
+            panic!("Container {} is not mounted", key)
         }
-        Ok(())
     }
 
-    pub async fn start(&mut self, name: &str) -> result::Result<(), Error> {
-        // Setup set of available resources
-        let resources = self
-            .resources
-            .values()
-            .map(|container| container.manifest.name.clone())
-            .collect::<HashSet<Name>>();
-
-        // Look for app
-        let app = if let Some(app) = self.applications.get_mut(name) {
-            app
-        } else {
-            return Err(Error::ApplicationNotFound);
-        };
+    pub(super) async fn start(&mut self, key: &ContainerKey) -> Result<(), Error> {
+        info!("Trying to start {}", key);
 
-        // Check if application is already running
-        if app.process.is_some() {
-            warn!("Application {} is already running", app.manifest().name);
-            return Err(Error::ApplicationRunning(app.manifest().name.clone()));
-        }
+        let mut mounted = Vec::new();
 
-        // Check if app is a resource container that cannot be started
-        if app.container.is_resource() {
-            warn!("Cannot start resource containers ({})", app);
-            return Err(Error::ApplicationNotFound);
-        }
+        let manifest = if let Some(container) = self.containers.get(key) {
+            // Check if the container is not a resouce
+            if container.manifest.init.is_none() {
+                warn!("Container {} is a resource", key);
+                return Err(Error::UnknownApplication(key.clone()));
+            }
+
+            // Check if the container is already started
+            if container.process.is_some() {
+                warn!("Application {} is already running", key);
+                return Err(Error::ApplicationStarted(key.clone()));
+            }
 
-        // Check for all required resources
-        for mount in app.container.manifest.mounts.values() {
-            if let Mount::Resource { name, .. } = mount {
-                if !resources.contains(name) {
-                    return Err(Error::MissingResource(name.clone()));
+            container.manifest.clone()
+        } else if self.npks.contains_key(key) {
+            let container = self.mount(&key).await?;
+            let manifest = container.manifest.clone();
+            mounted.push(key.clone());
+            self.containers.insert(key.clone(), container);
+            manifest
+        } else {
+            warn!("Unknown application {}", key);
+            return Err(Error::UnknownApplication(key.clone()));
+        };
+
+        // Find to be mounted resources
+        for key in manifest.mounts.values().filter_map(|m| match m {
```

Comment:
Here we mount a container and then check the needed resources and need to roll-back when something is missing. Would be better to first check if all resources are available and can be mounted, then try to mount the container

Reply:
That's a optimisation.
The mounts have to be paralleled in future anyway which is an open todo:
- assemble a list of npks to be mounted (including the "root" container)
- mount all of those containers in parallel (today it's sequential).

Reply:
yes, parallelizing will be good, but still better in this way:
* assemble a list of npks to be mounted *including* necessary resources
* mount needed resource containers
* mount *only* those application containers that can be mounted
that will have the advantage that we do not do unnecessary work (roll-back)

Reply:
Done in `8835f8a`.